lib/data/file/backend.js

```diff
@@ -43,31 +43,39 @@ export const backend = {
                 log.error('error opening filePath', { error: err });
                 return callback(errors.InternalError);
             }
-            const fileStream = fs.createWriteStream(filePath, { fd });
-
-            request.resume();
-            request.pipe(fileStream, { end: false }).on('error', err => {
-                log.error('error streaming data from request on write',
-                    { error: err });
-                return callback(errors.InternalError);
-            });
-            request.on('error', err => {
-                log.error('error streaming data from request on read',
-                    { error: err });
-                // close fileStream
-                return fs.close(fd, () => callback(errors.InternalError));
-            }).on('end', () => {
+            // disable autoClose so that we can close(fd) only after
+            // fsync() has been called
+            const fileStream = fs.createWriteStream(filePath,
+                                                    { fd,
+                                                      autoClose: false });
+            fileStream.on('finish', () => {
                 fs.fsync(fd, err => {
-                    fileStream.end();
+                    fs.close(fd);
                     if (err) {
                         log.error('error streaming data from request on fsync',
-                            { error: err });
+                                  { error: err });
                         return callback(errors.InternalError);
                     }
                     log.debug('finished writing data', { key });
                     return callback(null, key);
                 });
                 return undefined;
+            }).on('error', err => {
```

Comment:
I think we should remove the new file in such case, indeed to avoid making orphans if the input stream is interrupted. Though not related to this PR, so I will do it as part of the current refactoring.